src/decoder/lattice-incremental-decoder.cc

```diff
@@ -0,0 +1,1295 @@
+  if (not_first_chunk)
+    state_offset--; // since we do not append initial state in the first chunk
+
+  // A map between state label and the arc position (index)
+  // the incoming states of these arcs are initial states of the chunk
+  // and the olabel of these arcs are the key of this map (state label)
+  // The arc position are obtained from ArcIterator corresponding to the state
+  unordered_map<int32, size_t> initial_arc_map;
+  initial_arc_map.reserve(std::min((int32)1e5, config_.max_active));
+  for (StateIterator<CompactLattice> siter(clat); !siter.Done(); siter.Next()) {
+    auto s = siter.Value();
+    StateId state_appended = -1;
+    // We do not copy initial state, which exists except the first chunk
+    if (!not_first_chunk || s != 0) {
+      state_appended = s + state_offset;
+      KALDI_ASSERT(state_appended == olat->AddState());
+      olat->SetFinal(state_appended, clat.Final(s));
+    }
+
+    for (ArcIterator<CompactLattice> aiter(clat, s); !aiter.Done(); aiter.Next()) {
+      const auto &arc = aiter.Value();
+      // We do not copy initial arcs, which exists except the first chunk.
+      // These arcs will be taken care later in step 3.2
+      if (!not_first_chunk || s != 0) {
+        CompactLatticeArc arc_appended(arc);
+        arc_appended.nextstate += state_offset;
+        olat->AddArc(state_appended, arc_appended);
+      }
+      // Process state labels, which will be used in step 3.2
+      if (arc.olabel > config_.max_word_id) { // initial_arc
+        if (s == 0) { // record initial_arc in this chunk, we will use it right now
+          initial_arc_map[arc.olabel] = aiter.Position();
+          if (last_get_lattice_frame_ != 0) // Erase since we are not interested
+            initial_state_in_chunk_.erase(arc.olabel);
+        } else { // final_arc
+          // record final_arc in this chunk for the step 3.2 in the next call
+          KALDI_ASSERT(clat.Final(arc.nextstate) != CompactLatticeWeight::Zero());
+          final_arc_list_.push_back(
+              pair<int32, size_t>(state_appended, aiter.Position()));
+        }
+      }
+    }
+  }
+  // sanity check, remove it later
+  // KALDI_ASSERT(initial_state_in_chunk_.size() == 0);
+
+  // step 3.2: connect the states between two chunks
+  if (last_get_lattice_frame_ != 0) {
+    KALDI_ASSERT(final_arc_list_prev_.size());
+    StateId prev_final_state = -1;
+    for (auto &i : final_arc_list_prev_) {
+      MutableArcIterator<CompactLattice> aiter(olat, i.first);
+      aiter.Seek(i.second);
+      // Obtain the appended final arcs in the last chunk
+      auto &arc_appended = aiter.Value();
+      // Find out whether its corresponding Token still exists in the begin
+      // of this chunk. If not, it is pruned by PruneActiveTokens()
+      auto r = initial_arc_map.find(arc_appended.olabel);
+      if (r != initial_arc_map.end()) {
+        ArcIterator<CompactLattice> aiter_unappended(clat, 0); // initial state
```

Comment:
unappeneded is not very helpful, IMO.  Maybe you could call the two arcs that you are going to append arc1 and arc2 and then have just 'arc_appended' be the appended one.